### PR TITLE
Authentication pull request

### PR DESCRIPTION
**Pull request by shan1322**
PR to solve #5.

### CODE REVIEW

File: controllers/auth.controller.js [hunk 1]
```diff
@@ -0,0 +1,57 @@
+const passport = require('passport')
+const User = require('../models/user.model')
+
+exports.index = (req, res) => {
```
Comment (the-lost-explorer) [inline controllers/auth.controller.js]: Remove this function.

File: controllers/auth.controller.js [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    res.render('../views/welcome', {user: req.user})
+}
+// This function is used to fo loging purpose(Authentication)
+exports.login = (req, res) => {
+    if (req.user)
+        res.status(403).send({"message": "Already logged in. Logout to continue."})
+    else
+        passport.authenticate('local')(req, res, function(){
+            res.redirect('/')
```
Comment (the-lost-explorer) [inline controllers/auth.controller.js]: @shubham1172 Seems doubtful to me. Should redirect to some other endpoint.

File: controllers/auth.controller.js [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+// This function is called for signing a new user
+exports.signup = (req, res) => {
+    if (req.user) 
+        res.status(403).send({"message": "Cannot sign up. Logout to continue."})
+    else {
+        var user = new User({username: req.body.username,user_type: req.body.user_type})
+        
+        User.register(user,req.body.password, function (err, user) {
+                if (err) {
+                //  console.log("Error registering user", err)
```
Comment (the-lost-explorer) [inline controllers/auth.controller.js]: Remove this log comment.
```suggestion
                
```

File: controllers/auth.controller.js [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+                    res.status(304).send({"message": "A user with the given username is already registered."})
+                } else {
+                    console.log(`Registered user - ${user.username}.`)
```
Comment (the-lost-explorer) [inline controllers/auth.controller.js]: Remove this log statement.
```suggestion
                    
```

File: controllers/auth.controller.js [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+                    passport.authenticate('local')(req, res, function(){
+                        console.log('Done')
```
Comment (the-lost-explorer) [inline controllers/auth.controller.js]: Remove this log statement.
```suggestion
                        
```

File: middlewares/auth.middleware.js [hunk 6]
```diff
@@ -0,0 +1,40 @@
+exports.isAuthenticated = (req, res, next) => {
+    if (req.user) {
+        return next()
+    } else {
+        res.render('../views/index')
```
Comment (the-lost-explorer) [inline middlewares/auth.middleware.js]: Doubtful. @shubham1172

File: controllers/auth.controller.js (hunk 2 above)
Comment (shubham1172) [reply controllers/auth.controller.js]: It's fine, '/' should be the home page only. Middleware will handle the login result.

File: controllers/auth.controller.js [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+// This function is used to fo loging purpose(Authentication)
+exports.login = (req, res) => {
+    if (req.user)
+        res.status(403).send({"message": "Already logged in. Logout to continue."})
+    else
+        passport.authenticate('local')(req, res, function(){
+            res.redirect('/')
+        })
+}
+// This function called when unauthenticate user trys to login
```
Comment (shubham1172) [inline controllers/auth.controller.js]: Check comment.

File: controllers/auth.controller.js [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+exports.login_page = (req, res) => {
```
Comment (shubham1172) [inline controllers/auth.controller.js]: Why is this function required? Don't serve any views now.

File: controllers/auth.controller.js [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+                        res.redirect('/')
+                    })
+                }      
+        })
+    }
+
+}
+
+// This is used when user directly tries to access signup page
+exports.signup_page = (req, res) => {
```
Comment (the-lost-explorer) [inline controllers/auth.controller.js]: Remove this function

File: controllers/auth.controller.js [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+        passport.authenticate('local')(req, res, function(){
+            res.redirect('/')
+        })
+}
+// This function called when unauthenticate user trys to login
+exports.login_page = (req, res) => {
+    res.render('../views/login', {error: req.query.error})
+}
+
+exports.admin=(req,res) => {
```
Comment (the-lost-explorer) [inline controllers/auth.controller.js]: What does this function do?

File: middlewares/auth.middleware.js [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+
+exports.isNotAuthenticated = (req, res, next) => {
+    if (req.user) {
+        res.redirect('/')
+    } else {
+        return next()
+    }
+}
+
+exports.isadmin=(req,res,next) =>{
```
Comment (shubham1172) [inline middlewares/auth.middleware.js]: Check naming convention. Should be `is_authenticated`, `is_admin`, etc. I'd suggest removing `is_authenticated` and checking the everything here. Why is `is_not_authenticated` a part?

File: middlewares/auth.middleware.js [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+    if(req.user.user_type==2) {
+        return next()
+    } else {
+        res.status(403).send({"message": "The user is unauthorized."})
+    }
+}
+
+exports.iseb=(req,res,next)=> {
+    if(req.user.user_type==1 || req.user.user_type==2) {
```
Comment (shubham1172) [inline middlewares/auth.middleware.js]: We don't want `admin` to access `eb` views. Functionality can be forwarded internally without using the endpoint. So we should not allow admin to pass through this.

File: models/user.model.js [hunk 13]
```diff
@@ -0,0 +1,13 @@
+var mongoose = require('mongoose')
+var Schema = mongoose.Schema
+var passportLocalMongoose = require('passport-local-mongoose')
+
+var UserSchema = new Schema({
+    user_type: Number,
+    username: String,
+    password: String,
```
Comment (shubham1172) [inline models/user.model.js]: Refer the [wiki](https://github.com/VITCMUN/munverse/wiki/Data-Models#user) for the right schema.

**Pull request merged**